### PR TITLE
Review 8

### PR DESCRIPTION
**Pull request by Kraktun**
Review should focus on AmountComparator.java ~~or DataAnalyzer.java (only methods with tag @ author Michelon)~~.

### CODE REVIEW

File: ocr/src/main/java/com/ing/software/ocr/AmountComparator.java [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+                productsSum = analyzeAmount(s);
+            ++index;
+        }
+        if (productsSum != null)
+            OcrUtils.log(3, "analyzePrices", "List of prices, first value is: " + productsSum.toString());
+        else
+            OcrUtils.log(3, "analyzePrices", "List of prices, no value found");
+        while (index < possiblePrices.size() && possiblePrices.get(index).getPercentage() > 0) {
+            String s = possiblePrices.get(index).getText().getDetection();
+            if (OcrUtils.isPossibleNumber(s)) {
```
Comment (cristianlazzarin) [inline ocr/src/main/java/com/ing/software/ocr/AmountComparator.java]: String s is already defined and used at the same mode; You should define it at the top of the method and use always that

File: ocr/src/main/java/com/ing/software/ocr/AmountComparator.java [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+                //Accept the value
+                flagHasSubtotal(possibleSubTotal);
+            }
+            //now we may have the same value of decodedAmount, or its double (=*2)
+            if (decodedAmount.compareTo(productsSum) == 0) {
+                OcrUtils.log(3, "analyzePrices", "List of prices equals decoded amount");
+                flagHasPriceList(productsSum);
+            } else if (decodedAmount.compareTo(productsSum.divide(new BigDecimal(2).setScale(2, RoundingMode.HALF_UP), RoundingMode.HALF_UP)) == 0) {
+                OcrUtils.log(3, "analyzePrices", "List of prices/2 equals decoded amount");
+                flagHasPriceList(productsSum.divide(new BigDecimal(2).setScale(2, RoundingMode.HALF_UP), RoundingMode.HALF_UP));
```
Comment (cristianlazzarin) [inline ocr/src/main/java/com/ing/software/ocr/AmountComparator.java]: I think it's difficult to read method with a lot of nested parameters, what about to calcolate those in a temporary variable and then use the method?

File: ocr/src/main/java/com/ing/software/ocr/AmountComparator.java [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+            } else if (OcrUtils.findSubstring(productsSum.toString(), decodedAmount.toString()) <= maxDistance) {
+                //It's acceptable a distance of maxDistance from target
+                flagHasPriceList(productsSum);
+                OcrUtils.log(3, "analyzePrices", "List of prices diffs from decoded amount by at most " + maxDistance);
+            } else if (OcrUtils.findSubstring(productsSum.divide(new BigDecimal(2).setScale(2, RoundingMode.HALF_UP), RoundingMode.HALF_UP).toString(), decodedAmount.toString()) <= maxDistance) {
```
Comment (cristianlazzarin) [inline ocr/src/main/java/com/ing/software/ocr/AmountComparator.java]: see comment at 242

File: ocr/src/main/java/com/ing/software/ocr/AmountComparator.java [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+                OcrUtils.log(3, "analyzePrices", "List of prices/2 diffs from decoded amount by at most " + maxDistance);
+                flagHasPriceList(productsSum.divide(new BigDecimal(2).setScale(2, RoundingMode.HALF_UP), RoundingMode.HALF_UP));
```
Comment (cristianlazzarin) [inline ocr/src/main/java/com/ing/software/ocr/AmountComparator.java]: see comment at 242

**Comment by Kraktun** [comment]
I'm going to close this to avoid conflicts with other pr and simplify updates.

File: ocr/src/main/java/com/ing/software/ocr/AmountComparator.java [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+    /**
+     * @author Michelon
+     * @date 10-12-17
+     * Analyze a list of possible prices looking for cash (='contante') and change (='resto').
+     * If found update this object accordingly
+     * todo: handle tips and explicit taxes
+     * @param possiblePrices list of possible prices. Not null.
+     */
+    void analyzeTotals(@NonNull List<RawGridResult> possiblePrices) {
+        BigDecimal decodedAmount = getAmount();
```
Comment (cristianlazzarin) [inline ocr/src/main/java/com/ing/software/ocr/AmountComparator.java]: see comments about string s and nested parameters

File: ocr/src/main/java/com/ing/software/ocr/AmountComparator.java [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+     * @param amount  RawText containing possible amount. Not null.
+     * @param product RawText containing possible product price. Not null.
+     * @return true if product is in the same column (extended by percentage) as amount
+     */
+    private static boolean isProductPrice(@NonNull RawText amount, @NonNull RawText product) {
+        int percentage = 50;
+        RectF extendedRect = OcrUtils.partialExtendWidthRect(amount.getRect(), percentage);
+        RectF productRect = product.getRect();
+        return extendedRect.left < productRect.left && extendedRect.right > productRect.right;
+    }
```
Comment (cristianlazzarin) [inline ocr/src/main/java/com/ing/software/ocr/AmountComparator.java]: I think ``` return( extendedRect.left < productRect.left) && (extendedRect.right > productRect.right;)``` is easier to be read

**Pull request closed without merging**